ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/cache/CachedBonsaiWorldView.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright Hyperledger Besu contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+package org.hyperledger.besu.ethereum.bonsai.cache;
+
+import org.hyperledger.besu.datatypes.Hash;
+import org.hyperledger.besu.ethereum.bonsai.storage.BonsaiWorldStateKeyValueStorage;
+import org.hyperledger.besu.ethereum.bonsai.storage.BonsaiWorldStateKeyValueStorage.BonsaiStorageSubscriber;
+import org.hyperledger.besu.ethereum.core.BlockHeader;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class CachedBonsaiWorldView implements BonsaiStorageSubscriber {
+  private BonsaiWorldStateKeyValueStorage worldStateStorage;
+  private final BlockHeader blockHeader;
+  private long worldViewSubscriberId;
+  private static final Logger LOG = LoggerFactory.getLogger(CachedBonsaiWorldView.class);
+
+  public CachedBonsaiWorldView(
+      final BlockHeader blockHeader, final BonsaiWorldStateKeyValueStorage worldView) {
+    this.blockHeader = blockHeader;
+    this.worldStateStorage = worldView;
+    this.worldViewSubscriberId = worldStateStorage.subscribe(this);
+  }
+
+  public BonsaiWorldStateKeyValueStorage getWorldStateStorage() {
+    return worldStateStorage;
+  }
+
+  public long getBlockNumber() {
+    return blockHeader.getNumber();
+  }
+
+  public Hash getBlockHash() {
+    return blockHeader.getHash();
+  }
+
+  public synchronized void close() {
+    worldStateStorage.unSubscribe(this.worldViewSubscriberId);
+    try {
+      worldStateStorage.close();
+    } catch (final Exception e) {
+      LOG.warn("Failed to close worldstate storage for block " + blockHeader.toLogString(), e);
+    }
+  }
+
+  public synchronized void updateWorldStateStorage(
```

Comment:
Reason is/was because there is an implied chain of "parent" storages.  If we replaced the object in the cache we would have a "ripple" effect where we have to update all cache entries that represent descendents.  We wanted to minimize the need for locking and potential for concurrency issues.  

If we just replace the storage itself, there is no ripple and it can be a fairly cheap synchronized operation.